core/blockchain.go

```diff
@@ -355,6 +356,21 @@ func (bc *BlockChain) GetVMConfig() *vm.Config {
 	return &bc.vmConfig
 }
 
+// NewEVMCaller retuns the EVMCaller factory of the blockchain
+func (bc *BlockChain) NewEVMCaller(header *types.Header, state vm.StateDB) vm.EVMCaller {
+	return vmcontext.NewEVMCaller(bc, header, state)
+}
+
+// NewEVMCaller retuns the EVMCaller factory of the blockchain
+func (bc *BlockChain) NewCurrentHeadEVMCaller() (vm.EVMCaller, error) {
+	header := bc.CurrentHeader()
+	state, err := bc.StateAt(header.Root)
+	if err != nil {
+		return nil, err
+	}
+	return vmcontext.NewEVMCaller(bc, header, state), nil
+}
+
```

Comment:
I would consider not adding these methods to BlockChain. Given that long term we want to get rid of the object adding methods on it will add to the work we need to do in the future. Also having these methods on blockchain may encourage people to pass blockchain when really they only need an EVMCaller thus increasing coupling ( because blockchain has so much other stuff in it)

Reply:
I considered this; but i was convinced to go fro the lesser of evils. Blockchain is generally available through the code; and it's alreayd used to get other things like the `VMConfig`. You'll notice that i didn't had to do much "object passing" so that places where an evmCaller was required were able to fetch one.

Initially i put the factory on `Ethereum` & `LightEthereum` service; but that proved hard, since not many objects have reference to them, and it would imply passing a new object everywhere, while blockchain was already there.

I agree it's not the best, that changing that involves a much deeper refactor, that i would prefer to leave for a subsequent refactor.

Reply:
So yeah, if you want to get to cleaner code at a certain point you have to do the object passing, so either we do it now or we do it later. But adding methods like this is backing us into a corner and will increase the chances of encountering circular dependency problems. So if we continue to make compromises like this, eventually we will get stuck, so lets schedule the refactor as soon as possible.

Reply:
this PR is just too big as it is, so doing a architectural change so everything lives in another place, not the blockchian is to me work for another PR. Still, i'm not convinced that would be a great idea thinking on tradeoff valueObtained vs upstreamMergeComplexity. But we can chat about it later.